### PR TITLE
Support for Atcoder

### PR DESCRIPTION
**Pull request by s-i-d-d-i-s**
So, the workflow is pretty much the same.

Instead of Codeforces API i use, Kenokooo API for Atcoder to fetch all submissions, and then use BeautifulSoup to Scrape it.

Some of the changes i did
1. Added a `AtcoderWorkflow` which does the same job as of `CodeforcesWorkflow`
2. the `get_user_submissions` has an extra key now, which is `"platform"` because Readme can't be generated for all platforms in the same way, for e.g Atcoder lacks a Tag system for problems

Some more things like arguments etc, you can review them as well

Here is a [sample repo ](https://github.com/s-i-d-d-i-s/Atcoder-Test)

### CODE REVIEW

**Comment by nileshsah** [comment]
This is just amazing @s-i-d-d-i-s! Thank you so so much for working over it! Taking a look at the proposed changes :)

File: harwest/lib/utils/submissions.py [hunk 1]
```diff
@@ -34,11 +34,17 @@ def __generate_readme(self, submissions):
         continue
       problems.add(submission['problem_url'])
       row = str(index) + " | "
-      row += '[{problem_index} - {problem_name}]({problem_url}) | '.format(
-        problem_index=submission['problem_index'],
-        problem_name=submission['problem_name'],
-        problem_url=submission['problem_url']
-      )
+      if submission['platform'] == 'atcoder':
```
Comment (nileshsah) [inline harwest/lib/utils/submissions.py:37]: Users who might already have used harwest before to set up their repository will have the `platform` key missing in their submissions.js in which case it'll throw a KeyError. Let's just have a safe check of `'platform' in submission` before we compare it to atcoder. For cases wherein the key is missing, we can safely assume it to be from codeforces

File: harwest/harwest.py [hunk 2]
```diff
@@ -27,6 +28,19 @@ def build_argument_parser():
     'leetcode', help="Scrape solutions from the leetcode platform")
   lc_parser.set_defaults(func=leetcode)
 
+
```
Comment (nileshsah) [inline harwest/harwest.py:31]: I'm using the PEP8 styling convention for Python with the exception of using 2 spaces over 4 spaces. I've configured the linter to use the same on my IDE (Intellij) which helps a lot in quickly identifying and fixing formatting issues. Would recommend setting up the same in the IDE of your choice and fixing some of the formatting issues which might have cropped up in these changes.

File: harwest/lib/atcoder/workflow.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
+import os
+from harwest.lib.utils import config
+
+from harwest.lib.atcoder.client import AtcoderClient
+from harwest.lib.utils.repository import Repository
+from harwest.lib.utils.submissions import Submissions
+
+
+class AtcoderWorkflow:
```
Comment (nileshsah) [inline harwest/lib/atcoder/workflow.py:9]: Would completely agree with you that the behavior of this workflow is rather similar to that of the Codeforces one with some more minor changes in the flow leading to quite a bit of duplication of code. How about we extract out the common behavior into a separate file `lib/AbstractWorklfow.py` and make both the Codeforces and AtCoder workflow extend it?

File: harwest/lib/atcoder/workflow.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    print(text, end='\r')
+    return len(text)
+
+  def run(self, start_page_index=1):
+    print ("\U000026CF", "️Harvesting Atcoder (%s) Submissions to %s" %
+           (self.user_data['atcoder'], self.submissions_directory))
+    submission_index = start_page_index
+    try:
+      submissions = self.client.get_user_submissions(submission_index)
+      for index, submission in enumerate(submissions):
```
Comment (nileshsah) [inline harwest/lib/atcoder/workflow.py:91]: uhm, so if I understand the logic correctly, would we iterate over the entire submissions list every time we run `harwest atcoder`?

File: harwest/lib/utils/submissions.py (hunk 1 above)
Comment (s-i-d-d-i-s) [reply harwest/lib/utils/submissions.py:37]: yeah, that'll be better

File: harwest/harwest.py (hunk 2 above)
Comment (s-i-d-d-i-s) [reply harwest/harwest.py:31]: sure, i'll use sublime. i'll configure it accordingly

File: harwest/lib/atcoder/workflow.py (hunk 3 above)
Comment (s-i-d-d-i-s) [reply harwest/lib/atcoder/workflow.py:9]: yeah, i was thinking about that as well, it'll make it a bit easier to manage as well

File: harwest/lib/atcoder/workflow.py (hunk 4 above)
Comment (s-i-d-d-i-s) [reply harwest/lib/atcoder/workflow.py:91]: yeah, kenkoooo doesn't have a page parameter in its API so we get the whole submission as a result, most OJs won't give an official API so we won't able to use page parameter for all future OJs anyways

**Comment by nileshsah** [comment]
Hey @s-i-d-d-i-s! Thanks a ton again for contributing! The changes look all fine. I have some bandwidth to work on it today so I'll merge your changes and refactor things a bit to extract out a common workflow interface. I just might revert the changes for the Codeforces page limit at the moment to keep it in parity with the actual submissions page on Codeforces but will bring it back if more people complain, maybe with an additional `--all` argument. Hope that works?

**Comment by s-i-d-d-i-s** [comment]
yeah sure!

**Pull request merged**